File: backend/app/services/gitlab.py
from __future__ import annotations

import httpx

from ..schemas import GitLabConfig


class GitLabError(RuntimeError):
    pass


def _headers(config: GitLabConfig) -> dict:
    headers = {"Content-Type": "application/json"}
    if config.token:
        headers["PRIVATE-TOKEN"] = config.token
    return headers


def _base(config: GitLabConfig) -> str:
    return config.base_url.rstrip("/")


async def _get(config: GitLabConfig, path: str, params: dict | None = None) -> list | dict:
    async with httpx.AsyncClient(timeout=30) as client:
        resp = await client.get(
            f"{_base(config)}{path}", params=params, headers=_headers(config)
        )
    if resp.status_code != 200:
        raise GitLabError(f"GitLab 请求失败 ({resp.status_code}): {resp.text[:300]}")
    try:
        return resp.json()
    except ValueError as e:
        raise GitLabError(f"GitLab 响应解析失败: {resp.text[:300]}") from e


async def _put(config: GitLabConfig, path: str, payload: dict) -> dict:
    async with httpx.AsyncClient(timeout=30) as client:
        resp = await client.put(
            f"{_base(config)}{path}", json=payload, headers=_headers(config)
        )
    if resp.status_code != 200:
        raise GitLabError(f"GitLab 更新失败 ({resp.status_code}): {resp.text[:300]}")
    try:
        return resp.json()
    except ValueError as e:
        raise GitLabError(f"GitLab 响应解析失败: {resp.text[:300]}") from e


async def test_connection(config: GitLabConfig) -> str:
    try:
        data = await _get(config, "/api/v4/user")
        return f"连接成功（{data.get('username', '')}）"
    except Exception as e:  # noqa: BLE001
        raise RuntimeError(f"连接失败: {e}") from e


async def list_projects(config: GitLabConfig) -> list[dict]:
    data = await _get(config, "/api/v4/projects", {"membership": True, "per_page": 100})
    return [p for p in data if isinstance(p, dict)] if isinstance(data, list) else []


async def list_merge_requests(
    config: GitLabConfig,
    updated_after: str | None = None,
    updated_before: str | None = None,
    state: str = "all",
    per_page: int = 2000,
) -> list[dict]:
    params: dict = {
        "state": state,
        "scope": "all",
        "order_by": "updated_at",
        "per_page": per_page,
    }
    if config.username:
        params["author_username"] = config.username
    if updated_after:
        params["updated_after"] = updated_after
    if updated_before:
        params["updated_before"] = updated_before
    data = await _get(config, "/api/v4/merge_requests", params)
    return [m for m in data if isinstance(m, dict)] if isinstance(data, list) else []


async def get_merge_request(config: GitLabConfig, project_id: int, merge_request_iid: int) -> dict:
    data = await _get(config, f"/api/v4/projects/{project_id}/merge_requests/{merge_request_iid}")
    if not isinstance(data, dict):
        raise GitLabError("GitLab 返回了异常数据")
    return data


async def list_merge_request_commits(
    config: GitLabConfig, project_id: int, merge_request_iid: int
) -> list[dict]:
    data = await _get(
        config, f"/api/v4/projects/{project_id}/merge_requests/{merge_request_iid}/commits"
    )
    return [c for c in data if isinstance(c, dict)] if isinstance(data, list) else []


async def update_merge_request(
    config: GitLabConfig, project_id: int, merge_request_iid: int, title: str, description: str
) -> dict:
    return await _put(
        config,
        f"/api/v4/projects/{project_id}/merge_requests/{merge_request_iid}",
        {"title": title, "description": description},
    )

File: frontend/src-tauri/src/services/gitlab.rs
use anyhow::{anyhow, bail, Context, Result};
use serde_json::{json, Value};

use crate::models::GitLabConfig;

fn base_url(config: &GitLabConfig) -> String {
    config.base_url.trim_end_matches('/').to_string()
}

fn headers(config: &GitLabConfig) -> Vec<(String, String)> {
    let mut headers = vec![("Content-Type".to_string(), "application/json".to_string())];
    if !config.token.is_empty() {
        headers.push(("PRIVATE-TOKEN".to_string(), config.token.clone()));
    }
    headers
}

pub async fn get(
    client: &reqwest::Client,
    config: &GitLabConfig,
    path: &str,
    params: Vec<(&str, String)>,
) -> Result<Value> {
    let mut request = client
        .get(format!("{}{}", base_url(config), path))
        .timeout(std::time::Duration::from_secs(30))
        .query(&params);
    for (k, v) in headers(config) {
        request = request.header(k, v);
    }
    let resp = request.send().await.context("GitLab 请求失败")?;
    let status = resp.status();
    let text = resp.text().await.unwrap_or_default();
    if status != 200 {
        bail!("GitLab 请求失败 ({}): {}", status, truncate(&text, 300));
    }
    serde_json::from_str(&text)
        .map_err(|_| anyhow!("GitLab 响应解析失败: {}", truncate(&text, 300)))
}

pub async fn put(
    client: &reqwest::Client,
    config: &GitLabConfig,
    path: &str,
    payload: &Value,
) -> Result<Value> {
    let mut request = client
        .put(format!("{}{}", base_url(config), path))
        .timeout(std::time::Duration::from_secs(30))
        .json(payload);
    for (k, v) in headers(config) {
        request = request.header(k, v);
    }
    let resp = request.send().await.context("GitLab 请求失败")?;
    let status = resp.status();
    let text = resp.text().await.unwrap_or_default();
    if status != 200 {
        bail!("GitLab 更新失败 ({}): {}", status, truncate(&text, 300));
    }
    serde_json::from_str(&text)
        .map_err(|_| anyhow!("GitLab 响应解析失败: {}", truncate(&text, 300)))
}

fn as_list(data: Value) -> Vec<Value> {
    data.as_array()
        .cloned()
        .unwrap_or_default()
        .into_iter()
        .filter(|v| v.is_object())
        .collect()
}

pub async fn test_connection(client: &reqwest::Client, config: &GitLabConfig) -> Result<String> {
    let data = get(client, config, "/api/v4/user", vec![])
        .await
        .map_err(|e| anyhow!("连接失败: {e}"))?;
    let username = data
        .get("username")
        .and_then(|v| v.as_str())
        .unwrap_or("");
    Ok(format!("连接成功（{username}）"))
}

pub async fn list_projects(client: &reqwest::Client, config: &GitLabConfig) -> Result<Vec<Value>> {
    let data = get(
        client,
        config,
        "/api/v4/projects",
        vec![("membership", "true".into()), ("per_page", "100".into())],
    )
    .await?;
    Ok(as_list(data))
}

pub async fn list_merge_requests(
    client: &reqwest::Client,
    config: &GitLabConfig,
    updated_after: Option<&str>,
    updated_before: Option<&str>,
    state: &str,
    per_page: i64,
) -> Result<Vec<Value>> {
    let mut params: Vec<(&str, String)> = vec![
        ("state", state.to_string()),
        ("scope", "all".to_string()),
        ("order_by", "updated_at".to_string()),
        ("per_page", per_page.to_string()),
    ];
    if !config.username.is_empty() {
        params.push(("author_username", config.username.clone()));
    }
    if let Some(v) = updated_after {
        params.push(("updated_after", v.to_string()));
    }
    if let Some(v) = updated_before {
        params.push(("updated_before", v.to_string()));
    }
    let data = get(client, config, "/api/v4/merge_requests", params).await?;
    Ok(as_list(data))
}

pub async fn get_merge_request(
    client: &reqwest::Client,
    config: &GitLabConfig,
    project_id: i64,
    merge_request_iid: i64,
) -> Result<Value> {
    let data = get(
        client,
        config,
        &format!("/api/v4/projects/{project_id}/merge_requests/{merge_request_iid}"),
        vec![],
    )
    .await?;
    if !data.is_object() {
        bail!("GitLab 返回了异常数据");
    }
    Ok(data)
}

pub async fn list_merge_request_commits(
    client: &reqwest::Client,
    config: &GitLabConfig,
    project_id: i64,
    merge_request_iid: i64,
) -> Result<Vec<Value>> {
    let data = get(
        client,
        config,
        &format!("/api/v4/projects/{project_id}/merge_requests/{merge_request_iid}/commits"),
        vec![],
    )
    .await?;
    Ok(as_list(data))
}

pub async fn update_merge_request(
    client: &reqwest::Client,
    config: &GitLabConfig,
    project_id: i64,
    merge_request_iid: i64,
    title: &str,
    description: &str,
) -> Result<Value> {
    put(
        client,
        config,
        &format!("/api/v4/projects/{project_id}/merge_requests/{merge_request_iid}"),
        &json!({"title": title, "description": description}),
    )
    .await
}

fn truncate(s: &str, n: usize) -> String {
    let mut out: String = s.chars().take(n).collect();
    if s.chars().count() > n {
        out.push_str("...");
    }
    out
}
